[FLINK-4692] [tableApi] Add tumbling group-windows for batch tables

wuchong opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

AlexanderShoshin committed:
[FLINK-5159] [table] Add single-row broadcast nested-loop join.

- Support for arbitrary joins (cross, theta, equi) if one input has exactly one row.

This closes #2811.

StefanRRichter committed:
[FLINK-5146] Improved resource cleanup in RocksDB keyed state backend

StefanRRichter committed:
[FLINK-5173] Upgrade RocksDB dependency

wuchong committed:
[FLINK-4692] [tableApi] Add tumbling group-windows for batch tables

StephanEwen committed:
[FLINK-5169] [network] Add tests for channel consumption

This closes #2882.

StephanEwen committed:
[FLINK-5169] [network] Make consumption of InputChannels fair

uce committed:
[FLINK-5169] [network] Adjust tests to new consumer logic

Commit:
[FLINK-5109] [webfrontend] Fix invalid content-encoding

This closes #2898

beyond1920 committed:
[FLINK-5158] [table] Refactor BatchScan, DataSetScan, and BatchTableSourceScan.

This closes #2921.

Renkai committed:
[FLINK-5128] [kafka] Get Kafka partitions in FlinkKafkaProducer only if a partitioner is set

This closes #2893.

fhueske committed:
[FLINK-4676] [connectors] Merge batch and streaming connectors into common Maven module.

This closes #2897.

uce committed:
[FLINK-5201] [logging] Log loaded config properties on INFO level

uce committed:
[FLINK-5196] [logging] Don't log InputChannelDeploymentDescriptor

uce committed:
[FLINK-5198] [logging] Improve TaskState toString

uce committed:
[FLINK-5199] [logging] Improve logging in ZooKeeperSubmittedJobGraphStore

uce committed:
[FLINK-5207] [logging] Decrease HadoopFileSystem logging

uce committed:
[FLINK-5192] [logging] Improve log config templates

This closes #2899.

uce committed:
[FLINK-5194] [logging] Log heartbeats on TRACE level

greghogan committed:
[FLINK-5209] [webfrontend] Fix TaskManager metrics

Fixes a capitalization incompatibility when providing memory metrics to
the webfrontend. Numeric metrics now returned as numbers in the JSON
API. Non-byte numbers now localized in the webfrontend.

This closes #2902

StephanEwen committed:
[FLINK-5216] [checkpoints] 'Min Time Between Checkpoints' references timestamp after checkpoint

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetWindowAggregateRule.scala:
```diff
@@ -0,0 +1,75 @@
+
+class DataSetWindowAggregateRule
+  extends ConverterRule(
+      classOf[LogicalWindowAggregate],
```
缩进

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala:
```diff
@@ -84,7 +84,9 @@ case class ProcessingTimeTumblingGroupWindow(
 
   override def validate(tableEnv: TableEnvironment): ValidationResult =
-    super.validate(tableEnv).orElse(TumblingGroupWindow.validate(tableEnv, size))
+    super.validate(tableEnv)
```
建议不再这里做。到抽象类里面做。

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala:
```diff
@@ -173,7 +176,9 @@ case class ProcessingTimeSlidingGroupWindow(
-    super.validate(tableEnv).orElse(SlidingGroupWindow.validate(tableEnv, size, slide))
+    super.validate(tableEnv)
+      .orElse(SlidingGroupWindow.validate(tableEnv, size, slide))
+      .orElse(ProcessingTimeWindow.validate(tableEnv, size))
```
同上

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala:
```diff
@@ -196,7 +201,8 @@ case class EventTimeSlidingGroupWindow(
       .orElse(size match {
-        case Literal(_, RowIntervalTypeInfo.INTERVAL_ROWS) =>
+        case Literal(_, RowIntervalTypeInfo.INTERVAL_ROWS)
+          if tableEnv.isInstanceOf[StreamTableEnvironment] =>
```
同上

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala:
```diff
@@ -232,7 +238,9 @@ case class ProcessingTimeSessionGroupWindow(
-    super.validate(tableEnv).orElse(SessionGroupWindow.validate(tableEnv, gap))
+    super.validate(tableEnv)
+      .orElse(SessionGroupWindow.validate(tableEnv, gap))
+      .orElse(ProcessingTimeWindow.validate(tableEnv, gap))
```
同上

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala:
```diff
@@ -256,3 +264,14 @@ case class EventTimeSessionGroupWindow(
   override def toString: String = s"EventTimeSessionGroupWindow($name, $timeField, $gap)"
 }
+
+object ProcessingTimeWindow {
```
若可以在抽象类做，这里就不需要了。

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala:
```diff
@@ -0,0 +1,268 @@
+    * @return window size (with unit),
+    *         and time field position & type information if it is an event-time window
+    */
+  private def getWindowInformation(groupWindow: LogicalWindow)
```
需要明天讨论

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -84,14 +86,138 @@ object AggregateUtil {
+
+    // the addition one field is used to store time attribute
+    val (aggFieldIndexes, aggregates) = transformToAggregateFunctions(
+      namedAggregates.map(_.getKey), inputType, groupings.length + 1)
```
不知道加明确的名字好?还是这样好

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/WindowAggregateMapFunction.scala:
```diff
@@ -0,0 +1,97 @@
+    private val windowSize: Long,
+    private val isTimeWindow: Boolean,
+    @transient private val returnType: TypeInformation[Row])
+    extends RichMapFunction[Row, Row] with ResultTypeQueryable[Row] {
```
缩进

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala:
```diff
@@ -649,9 +649,6 @@ class Table(
     * @return A windowed table.
     */
   def window(groupWindow: GroupWindow): GroupWindowedTable = {
```
我想在这里检查是不是比较好！（该检查的都提前检查）

StephanEwen committed:
[hotfix] IOUtils.closeQuietly() closes absolutely quietly.

StephanEwen committed:
[FLINK-5218] [state backends] Add test that validates that Checkpoint Streams are eagerly closed on cancellation.

This is important for some stream implementations (such as HDFS) that do not properly
handle thread interruption.

StephanEwen committed:
[hotfix] Fix test instability in AbstractTaskManagerProcessFailureRecoveryTest

uce committed:
[FLINK-5169] [network] Fix spillable subpartition buffer count

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala:
```diff
@@ -0,0 +1,268 @@
+          namedAggregates,
+          namedProperties))
+  }
+
```
是不是应该有计算cost.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -84,14 +86,138 @@ object AggregateUtil {
+    groupings: Array[Int],
+    inputType: RelDataType): MapFunction[Any, Row] = {
+
+    // the addition one field is used to store time attribute
```
是否可以根据window类型创建不同的mapfunction

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -84,14 +86,138 @@ object AggregateUtil {
+    }
+
+    if (window.isTimeWindow) {
+      // time window
```
是否可以根据window类型创建不同的reduceGroupFunction.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -308,7 +434,7 @@ object AggregateUtil {
     * Return true if all aggregates can be partially computed. False otherwise.
     */
-  private[flink] def doAllSupportPartialAggregation(
+  def doAllSupportPartialAggregation(
```
为啥去掉private[flink]

wuchong replied:
`doAllSupportPartialAggregation` 以及 `createAggregateGroupReduceFunction`  都是该util提供给其他类使用的帮助函数，所以应该设置为public的可见性。`private[flink]` 应该用在该类是提供给用户的，而该方法不希望对用户可见的。但是 `AggregateUtil.scala` 是非面向用户的，所以不需要`private[flink]`。 如 `CodeGenerator` ， `CodeGenUtils` ， `UserDefinedFunctionUtils` 等等都是类似的。

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -84,14 +86,138 @@ object AggregateUtil {
     *
     */
-  private[flink] def createAggregateGroupReduceFunction(
+  def createAggregateGroupReduceFunction(
```
建议加上private[flink]

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala:
```diff
@@ -722,9 +719,6 @@ class GroupedTable(
     * @return A windowed table.
     */
   def window(groupWindow: GroupWindow): GroupWindowedTable = {
-    if (table.tableEnv.isInstanceOf[BatchTableEnvironment]) {
```
这里检查所有batch上支持的窗口类型是否更好些。该检查的都提前检查

fhueske committed:
[hotfix] [jdbc] Set flink-table dependency to optional to avoid pulling in Calcite if not needed.

uce committed:
[FLINK-5228] [network] Fix LocalInputChannel re-trigger request and release deadlock
